### PR TITLE
Adds findings in bucket level monitor 

### PR DESCRIPTION
**Pull request by eirsep**
This change surfaces out the list of documents that become a part of bucket level monitor aggregation buckets as findings for the monitor.

Signed-off-by: Surya Sashank Nistala <snistala@amazon.com>

*CheckList:*
[x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by sbcd90 — APPROVED** [review]

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 1]
```diff
@@ -287,6 +319,105 @@ object BucketLevelMonitorRunner : MonitorRunner() {
...
+                val bucketValues: Set<String> = triggerResult.aggregationResultBuckets.keys
+                val query = input.query
+                var fieldName = ""
+                var grouByFields = 0 // if number of fields used to group by > 1 we won't calculate findings
+                for (aggFactory in (query.aggregations() as AggregatorFactories.Builder).aggregatorFactories) {
+                    val sources = (aggFactory as CompositeAggregationBuilder).sources()
+                    for (source in sources) {
+                        if (grouByFields > 0) {
+                            return listOf()
+                        }
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:341]: Why do we need to return from here?

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 2]
```diff
@@ -287,6 +319,105 @@ object BucketLevelMonitorRunner : MonitorRunner() {
...
+                logger.debug("Findings: $findingStr")
+                if (shouldCreateFinding) {
+                    val indexRequest = IndexRequest(monitor.dataSources.findingsIndex)
+                        .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                        .source(findingStr, XContentType.JSON)
+                        .id(finding.id)
+                        .routing(finding.id)
+
+                    monitorCtx.client!!.suspendUntil<Client, IndexResponse> {
+                        monitorCtx.client!!.index(indexRequest, it)
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:411]: Are we indexing each finding document, one at a time? Wondering if we should use bulk request instead?

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt [hunk 3]
```diff
@@ -287,6 +319,105 @@ object BucketLevelMonitorRunner : MonitorRunner() {
...
+                    }
+                }
+
+                findings.add(finding.id)
+            }
+        }
+        return findings
```
Comment (getsaurabh02) [inline alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:418]: Do we need to return the findings from here, as indexing is already done?

File: alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt [hunk 4]
```diff
@@ -1321,6 +1323,71 @@ class MonitorRunnerServiceIT : AlertingRestTestCase() {
         )
     }
 
+    fun `test bucket-level monitor with findings enabled`() {
```
Comment (getsaurabh02) [inline alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt:1326]: Lets validate other scenarios as well:
1. Where findings will not be created - such as compound aggregation.
2. Findings index unavailable, or exception in index.

**Review by getsaurabh02 — COMMENTED** [review]
Thanks @eirsep . Overall approach looks good to me. Few comments/clarifications to consider.

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt (hunk 1 above)
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:341]: If there are more groupby fields then aggregation buckets would become tuples. We can develop iteratively and first begin by supporting only for a single field(buckets with single field aggregated on) to filter on.

File: alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt (hunk 3 above)
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/BucketLevelMonitorRunner.kt:418]: we are returning finding ids. finding ids list is populated in the alert.

**Review by sbcd90 — APPROVED** [review]

**Review by getsaurabh02 — APPROVED** [review]

**Pull request merged**